### ISSUE TITLE
Is it possible to improve the filters?

### ISSUE THREAD

**Issue by PoisonFrog** [Betterbird/thunderbird-patches#35]
My biggest issue with Thunderbird is how weak the filters are. For example, imagine I want to filter messages from Sarah, Julia or Emily that contain the word cat, dog, rabbit or hamster in the subject.

In most email clients I tried I can do that by creating a single filter, but in Thunderbird you cannot use "OR" so you have to choose between matching all the conditions or any of the conditions.

The ideal should be:
From: Sarah **OR** From: Julia **OR** From: Emily
**AND**
Subject contains: cat **OR** Subject contains: dog **OR** Subject contains: rabbit **OR** Subject contains: hamster

To make it work in Thunderbird I have to create multiple filters for one simple task.

[This feature has been requested many times during this past decade, but Mozilla devs never cared to implement it. Apparently it is possible to do it if you edit manually the filter files, but not on the GUI.](https://forums.mozillazine.org/viewtopic.php?f=39&t=2677919)

**Comment by Betterbird** [comment]
Interesting. I have heaps of "simple" filters. The backend logic can evaluate any logical term, so yes, editing the filter definition file will possibly do the trick.

Filters are up for a re-write in a future version (https://bugzilla.mozilla.org/show_bug.cgi?id=1574136), but work hasn't started yet.

Is there a ticket/bug in Bugzilla requesting the more sophisticated logical terms? If not, it would be worth filing such a bug or at least commenting on the one I mentioned, so it might get done in the rewrite.

**Comment by PoisonFrog** [comment]
It seems like we have to wait for Mozilla to switch filtering language to JavaScript then.

**Comment by Betterbird** [comment]
[Bug 1574136](https://bugzilla.mozilla.org/show_bug.cgi?id=1574136) is about switching the backend from JS to C++, the frontend, the filter editor is already implemented in JS/(X)HTML/XUL.

We're at Betterbird are looking at improvements that TB hasn't implemented in years or decades, so an improved filter editor appears to be a good idea, especially since the backend already supports more sophisticated logical expressions. The design issue is how to present editing of complex logical expressions to the user in the UI. The [advanced search in Bugzilla](https://bugzilla.mozilla.org/query.cgi?format=advanced) has a so-called "Custom Search" that offers AND, OR and NOT terms, and instead of using parenthesis, they have groups of the same logical connector:
![image](https://user-images.githubusercontent.com/86832723/156758104-5cfb4aac-b18a-4ee3-ad5a-d766113421cc.png)

Doing something like this in the filter editor would be quite complex, but one could instead offer - additionally to "match all" (AND) and "match any" (OR) - an "advanced" input where users can add their own search condition, as was suggested in the [Mozillazine article](https://forums.mozillazine.org/viewtopic.php?f=39&t=2677919), so as a mock-up:

![image](https://user-images.githubusercontent.com/86832723/156760062-a7fd46d7-aab9-47cb-bdc7-2a80d4a6326e.png)

What do you think?

**Comment by Betterbird** [comment]
We looked into this a bit more. Letting the user edit the search term as suggested two comments above doesn't seem to be a future-proof (and user-friendly) option after all. The "real" solution of adding grouping seems the way to go. See latest comment here:
https://bugzilla.mozilla.org/show_bug.cgi?id=297852#c65

**Comment by Betterbird** [comment]
@PoisonFrog, do you think you or someone else could drum up some financial incentive for this?

**Comment by PoisonFrog** [comment]
> @PoisonFrog, do you think you or someone else could drum up some financial incentive for this?

I would love to, but the reality is that I live in a third world country and since our currency is very weak compared to the dollar and our salaries are not that great to begin with, anything that I might be able to spend would not be a good incentive at all.

I understand that this is a feature that most people probably don't need and it must be a lot of work to implement something like this so if that doesn't make sense for your project right now you should drop it. I don't think I am able to provide meaningful support due to exchange rates, sorry.

**Comment by Betterbird** [comment]
Well, I wasn't saying that you need to pay for it yourself. There as FOSS funding sites like http://www.fossfactory.org/ or https://bountysource.com/. That said, someone has just inquired about the feature here again: https://bugzilla.mozilla.org/show_bug.cgi?id=297852#c66, so maybe the TB people will do something about it.

**Comment by Betterbird** [comment]
We've started working on it:
![image](https://user-images.githubusercontent.com/86832723/163650599-dade4802-29dd-4bd4-a5af-1e9eb0b4e451.png)

**Comment by RealRaven2000** [comment]
Great start! 

Here is a suggestion for improving readability, add operators to the start of the most relevant lines. Personally I find the "any / all" hard to read (the distinguishing word is after all only 10% of the whole, duplicated sentence **Match _any_ of the following** vs **Match _all_ of the following** - is there any meaningul way to shorten / clarify this after the first one?

Here are 2 variations using C-style and canonical boolean operators:

![image](https://user-images.githubusercontent.com/6314674/163714242-2944ba95-d186-47a5-9091-deffc3d09c20.png)
or
![image](https://user-images.githubusercontent.com/6314674/163714265-f5877be2-6d8a-46d1-8f91-4ad09dc8c0f3.png)

that together with indentation (in lieu of using actual brackets), would make the filter very easy to read, at least for my own eyes. (I omitted the very first && operator of each group for clarity, but personally I wouldn't even mind if it was shown on every line as well). Not sure if the layman would prefer this and whether they would like to use other operators such as  ⋀  ⋁ or maybe even & and | which would still be easy enough to understand for non-programmers.

**Comment by Betterbird** [comment]
Thanks for the comment, but I think we'll stay with the existing strings, which are similar to the native Mac UI: https://bugzilla.mozilla.org/show_bug.cgi?id=297852#c65

**Comment by RealRaven2000** [comment]
> Thanks for the comment, but I think we'll stay with the existing strings, which are similar to the native Mac UI: https://bugzilla.mozilla.org/show_bug.cgi?id=297852#c65

ok. although Mac seems actually a little superior using the dropdown for this. The spatial thing with 2 sentences jumping left and right seems messy and cluttered (also in the original Thunderbird UI). It's probably not easy to do this (the Mac version) in l10n though.

**Comment by Betterbird** [comment]
Please try our latest build: https://www.betterbird.eu/downloads/get.php?os=win&lang=en-US&version=latest

**Comment by PoisonFrog** [comment]
Can't see the changes in the Arch Linux AUR version yet. Also, I noticed a minor visual bug in Betterbird but I will create another issue for it.

**Comment by Betterbird** [comment]
Yes, the "latest build" is only provided for Windows.

**Comment by ThiloteE** [comment]
Thinking of trying this, but I am unsure if it would destroy my existing setup.

Will this work?

> You can install the same dot release of Thunderbird in parallel and switch between Betterbird and Thunderbird on the same profile without problems and without needing the -allow-downgrade command line switch

- source: https://www.betterbird.eu/faq/index.html

Will my old filters work with Betterbird out of the box? If I create new filters, will I be able to use them with Thunderbird? Even if not, what will happen, if I start Thunderbird and it detects these new filters? Will there be an error?

**Comment by ThiloteE** [comment]
Suggestion to rename this issue here to:

   "Fix for bugzilla issue 297852. Enable Boolean expressions in Filter UI (nested/grouped/mixed/combined/complex filters having AND and OR conditions)"

**Comment by Betterbird** [comment]
Yes, you can switch between TB and BB with no problems, unless, of course, you use special BB features. If you use the multi-line view, the subjects will be hidden and they don't come back if you go back to TB. You need to switch the subjects on in BB before switching back to TB.

We've just uploaded a new installer which you can get via https://www.betterbird.eu/downloads/get.php?os=win&lang=en-US&version=latest

There are a few incompatibilities now: BB implements expression grouping, and the [ ] are now used in msgFilterRules.dat to denote beginning and ending of so-called groups. So if you had something line `condition="OR (subject,contains,[List1]) OR (subject,contains,[list2])"`, that now needs to become `condition="OR (subject,contains,\"[List1]\") OR (subject,contains,\"[list2]\")"`. You need to inspect msgFilterRules.dat for that. We suggest to take a backup of msgFilterRules.dat.

If you create a complex expression, that will of course not work in TB. We haven't tested yet, how TB will react to the new `[ ]` syntax. At best, the rule won't be loaded, worse, the entire filter won't be loaded, or at worst, no filter at all will be loaded.

**Comment by ThiloteE** [comment]
Thank you for the instant reply and explanation! Will try it.

**Comment by Betterbird** [comment]
Further to comment https://github.com/Betterbird/thunderbird-patches/issues/35#issuecomment-1104040730. We've uploaded a new build that fixes a few issues and also introduces backward compatibility to TB msgFilterRules.dat so that unescaped closing `]` will still be accepted, so no manual conversion necessary. BB will write them out packaged into a string, but TB can already read that. So unless the new groups are used in BB, it's still 100% compatible with TB. How good is that? 😁

**Comment by Betterbird** [comment]
TB/BB new rule compatibility:

If you don't use the new feature, TB and BB remain 100% compatible.
If you do use the new feature, TB will destroy the rule that contains the complex search term syntax when opening the filter editor.

**Comment by RealRaven2000** [comment]
I just installed 91.10 - there still seems to be an issue with changing the operator for grouped items, I noticed this using "any" on both outer and inner groups:
![image](https://user-images.githubusercontent.com/6314674/171820170-16d3d1d6-73b0-4378-8916-4fd34181ac35.png)

when I change the top operator 
![image](https://user-images.githubusercontent.com/6314674/171820289-c68740f5-9fa4-47e9-a4fd-6d45bc892ac2.png)

and save (click ok) - both operators get changed internally:
![image](https://user-images.githubusercontent.com/6314674/171820347-5a7de39f-8bda-4b53-9f95-1d0201d4409c.png)


trying to correct this by switching the inner operator to "any":
![image](https://user-images.githubusercontent.com/6314674/171820441-5c88162d-515a-4104-8540-cc47c89b9a85.png)

after clicking ok, once I reopen the filter, both operators are switched bad to "all" again:
![image](https://user-images.githubusercontent.com/6314674/171820552-d3983b80-e1b1-4f03-a350-98ee0cbfebfc.png)

Here is the filter definition:
```
version="9"
logging="no"
name="[Gmail] » Youtube"
enabled="yes"
type="16"
action="Move to folder"
actionValue="imap://thunderbirddaily%40gmail.com@imap.gmail.com/[Gmail]/Youtube"
condition="OR (subject,contains,creator] AND [to,contains,noreply@youtube.com) OR (from,contains,noreply@youtube.com]"

```

**Comment by Betterbird** [comment]
Here it works, all combinations of AND/OR are saved and properly restored.

Something is wrong with your rule, see `condition="OR (subject,contains,creator]` - How did you manage to get the closing `]` into the first term? There is no group ending there.

I suggest to delete the filter and start again. If you can give us the exact steps how you got into this situation, we'll look into it, but as we said, we can't reproduce the issue.

**Comment by RealRaven2000** [comment]
You're right, this filter is invalid - again not obvious on first glance. I think it was my wrong understanding again - I must have added the group after generating a filter:

![image](https://user-images.githubusercontent.com/6314674/171867362-2766339d-306f-42e8-8ca9-1cf257f33bd0.png)

back then I expected this to move the following search terms into the new group. I think I only found the filter again and tried to correct it - again - not seeing that the "From + Too" search terms were not indented under the operator.

If I start with a number of email addresses (let's say 10 pieces, gathered from my assistant) which are "any, from, contains":
![image](https://user-images.githubusercontent.com/6314674/171869111-c53c2e50-ec72-4ada-849e-150f98a9aa4a.png)

...and then want to add a new condition "only if subject line contains abc" it's actually not possible to do this without removing all conditions. You have to start creating the groups first and then add conditions.... logically I think all conditions are already in a group and the user should be able to add "separate groups beside it" - this was how I was hoping the top "add group" button would work (rather than inserting a child group, I always expect it to generate a sibling group). There is no way to add a group at the top level (split the search list into 2 groups), I think that's probably the button I was looking for...

(A|B|C)  ==>  (X) & (A|B|C)

if there is any way to achieve this in the editor besides stating from scratch I would sure like to know. At least we can simplify a complex filter but we don't seem to be able to make a simple filter more complex.

PS: I need to figure out a way (maybe through the quickFilters extension) to split / move a number of search terms into a new group to make it really useful. (e..g for refining a filter with many "any" terms that needs to be combined with an extra "all" condition. at the moment this can only be done by carefully planning the grouping parts before populating them. Merging new conditions to existing filters from code will be a similar challenge - so far new search terms were simply added at the end assuming "any" operator - now it needs to be based on similar search conditions (e.g. from contains "X" within a group, add new "from contains "Y" to the group).

**Comment by Betterbird** [comment]
A closing `]` without an opening one was a bug caused by group deletion, this was fixed in 91.9.1. So we suggest to delete the faulty filter and start again. If you can still generate invalid filters with the current version, please file a new ticket giving the exact steps.

> You have to start creating the groups first and then add conditions

Yes, if you create simple conditions at the top level, all groups created at top level will precede those simple conditions. So if you want the groups to follow the simple conditions, create them first.

This ticked has become rather confusing. If there are still bugs, please file a new issue giving the exact steps of how to get to the situation. See issue #60 for an example of causing a lot of confusion when the steps aren't given. Finally a video was attached that clearly shows a user error.
